src/differentiation.jl

```diff
@@ -25,57 +25,77 @@ differentiate(p, (x, y)) # should return [6xy+1, 3x^2+1]
 differentiate(p::APL, v::AbstractVariable) = polynomial(differentiate.(terms(p), v), SortedState())
-
 differentiate(p::RationalPoly, v::AbstractVariable) = (differentiate(p.num, v) * p.den - p.num * differentiate(p.den, v)) / p.den^2
 
 const ARPL = Union{APL, RationalPoly}
 
-_vec_diff_promote_op(::Type{PT}, ::AbstractVector{VT}) where {PT, VT} = _diff_promote_op(PT, VT)
-_vec_diff_promote_op(::Type{PT}, ::NTuple{N, VT}) where {PT, N, VT}   = _diff_promote_op(PT, VT)
-_vec_diff_promote_op(::Type{PT}, ::VT, xs...) where {PT, VT}          = _diff_promote_op(PT, VT)
-_vec_diff_promote_op(::Type{PT}, xs::Tuple) where PT = _vec_diff_promote_op(PT, xs...)
-
-# even if I annotate with ::Array{_diff_promote_op(T, PolyVar{C}), N+1}, it cannot detect the type since it seems to be unable to determine the dimension N+1 :(
-function differentiate(ps::AbstractArray{PT, N}, xs::Union{AbstractArray, Tuple}) where {N, PT<:ARPL}
-    qs = Array{_vec_diff_promote_op(PT, xs), N+1}(length(xs), size(ps)...)
-    for (i, x) in enumerate(xs)
-        for j in linearindices(ps)
-            J = ind2sub(ps, j)
-            qs[i, J...] = differentiate(ps[J...], x)
-        end
-    end
-    qs
+function differentiate(ps::AbstractArray{PT}, xs::AbstractArray) where {PT <: ARPL}
+    differentiate.(reshape(ps, (1, size(ps)...)), reshape(xs, :))
+end
+
+function differentiate(ps::AbstractArray{PT}, xs::Tuple) where {PT <: ARPL}
+    differentiate.(reshape(ps, (1, size(ps)...)), xs)
 end
 
+
+# TODO: this signature is probably too wide and creates the potential
+# for stack overflows
 differentiate(p::ARPL, xs) = [differentiate(p, x) for x in xs]
 
 # differentiate(p, [x, y]) with TypedPolynomials promote x to a Monomial
 differentiate(p::ARPL, m::AbstractMonomial) = differentiate(p, variable(m))
 
-# In Julia v0.5, Base.promote_op returns Any for PolyVar, Monomial and MatPolynomial
-# Even on Julia v0.6 and Polynomial, Base.promote_op returns Any...
-_diff_promote_op(::Type{PT}, ::Type{VT}) where {PT, VT} = Base.promote_op(differentiate, PT, VT)
-_diff_promote_op(::Type{MT}, ::Type{<:AbstractVariable}) where {MT<:AbstractMonomialLike} = termtype(MT, Int)
-
-function differentiate(p, x, deg::Int)
+# The `R` argument indicates a desired result type. We use this in order
+# to attempt to preserve type-stability even though the value of `deg` cannot
+# be known at compile time. For scalar `p` and `x`, we set R to be the type
+# of differentiate(p, x) to give a stable result type regardless of `deg`. For
+# vectors p and/or x this is impossible (since differentiate may return an array),
+# so we just set `R` to `Any`
+function _differentiate_recursive(p, x, deg::Int, ::Type{R}) where {R}
     if deg < 0
         throw(DomainError())
     elseif deg == 0
-        # Need the conversion with promote_op to be type stable for PolyVar, Monomial and MatPolynomial
-        return convert(_diff_promote_op(typeof(p), typeof(x)), p)
+        return convert(R, p)::R
```

Comment:
Wouldn't return type annotation do exactly that ? e.g.
```julia
function _differentiate_recursive(p, x, deg::Int, ::Type{R})::R where {R}
```

Reply:
Oh, good point. I'll do that.